# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

as well as an

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [1]:
#Imports:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [2]:
#Code:
sat_2017 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/sat_2017.csv')
act_2017 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/act_2017.csv')

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [4]:
#Code:
sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [5]:
act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

   Composite  
0       21.0  
1       19.2  
2       19.8  
3       19.7  
4       19.4  
5       22.8  
6       20.8  
7       25.2  
8       24.1  
9       24.2

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

**Answer:** First we will start with the sat_2017 dataset. This dataset contains 51 rows for all 50 states and the District of Columbia. There are 5 columns in this data set: 'State', 'Participation', Evidence-Based Reading and Writing', 'Math', and 'Total.' In sat_2017 there are no null values which means the dataset is complete. The average scores across all 50 states and the District of Columbia for Evidence-Based Reading and Writing is a score of 569. The average scores across all 50 states and the District of Columbia for Math is a score of 547. The average scores across all 50 states and the District of Columbia for the toal is a score of 1126.

The act_2017 dataset contains 52 rows. This includes 50 rows for all 50 states, a row for the District of Columbia and a row for the nation as a whole. There are 7 columns in this dataset: 'State', 'Participation', 'English', 'Math', 'Reading', 'Science', 'Composite.'In act_2017 there are no null values which means the dataset is complete. However in the 'Composite' column Wyoming has a value of 20.2x which is not like the other values so this was most likely an error which occured in data entry. In the 'Science' column Maryland has a value of 2.3 which is not possible. The average score for English was 20.9, the average socre for Math was 21.1, the average score for reading ws 22.0, the average score for science was 21.0. 

#### 4a. Does the data look complete? 

**Answer:**
No, the data is not complete. There are a few errors in both datasets. 


The sat_2017 dataset is not complete because the value for Maryland in the 'Math' column is 52 which is impossible. The act_2017 dataset is not complete because in the 'Composite' column Wyoming has a value of 20.2x which makes the 'Composite' column datatype an object instead of an integer. Also, the value for Maryland in the 'Science' column is 2.3 which is not possible.

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

**Answer:**

sat_2017
* **ERROR** Accoring to https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/ the value for Maryland in the 'Math' column should be 524.
* The minimum for Evidence-Base Reading and riding was 482. The minimum for the total was 950. The maximum for Math is 651. 
* The maximum for Evidence-Based Reading and Writing is 644. The maximum for total is 1295.

act_2017

* **ERROR** Accoring to https://www.act.org/content/dam/act/unsecured/documents/cccr2017/ACT_2017-Average_Scores_by_State.pdf the value for Maryland in the 'Science' column should be 23.3 not 2.3
* **ERROR** In the 'Composite' column Wyoming has a value of 20.2x. It should just be 20.2
* The minimum value in the 'Science' column is 2.3. The minimum for 'English' is 16.3. The minimum for 'Math' is 18.0. The minimum for 'Reading' is 18.1. 
* The maximum for 'English' is 25.5. The maximum for 'Math' is 25.3. The maximum for 'Reading' is 26. The maximum for 'Science' is 24.9.

In [7]:
sat_2017.head(1)

State Participation  Evidence-Based Reading and Writing  Math  Total
0  Alabama            5%                                 593   572   1165

In [6]:
act_2017.head(1)

State Participation  English  Math  Reading  Science  Composite
0  National           60%     20.3  20.7     21.4     21.0       21.0

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [8]:
# How to fix the errors I identified

# This fixes the error of Maryland in the column 'math_sat' Changes 52 to 524
sat_2017.at[20, 'Math'] = 524

# This fixes the error for Maryland in the column 'science_act' Changes 2.3 to 23.2
act_2017.at[21, 'Science']= 23.2

# This fixes the error for Wyoming in the column 'composite' Changes 20.2x to 20.2
act_2017.at[51, 'Composite']= 20.2

In [11]:
# We can see the updated values here for the sat_2017 dataset
sat_2017.loc[(sat_2017['State'] == 'Maryland')]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536   524   1060

In [12]:
# We can see the updated values here for the act_2017 dataset
act_2017.loc[(act_2017['State'] == 'Maryland')]

State Participation  English  Math  Reading  Science  Composite
21  Maryland           28%     23.3  23.1     24.2     23.2       23.6

In [13]:
# We can see the updated values here for the act_2017 dataset
act_2017.loc[(act_2017['State'] == 'Wyoming')]

State Participation  English  Math  Reading  Science  Composite
51  Wyoming          100%     19.4  19.8     20.8     20.6       20.2

#### 5. What are your data types? 
Display the data types of each feature. 

In [18]:
# Datatypes for sat_2017
sat_2017.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [19]:
# Datatypes for act_2017
act_2017.dtypes

State             object
Participation     object
English          float64
Math             float64
Reading          float64
Science          float64
Composite        float64
dtype: object

What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

**Answer:**

The 'Participation' columns should be a float datatype so we can work with it later on. Also, in the act_2017 dataset the 'Composite' column datatype was an object before I updated the value. Now it is the proper datatype.

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [27]:
#code

# This code makes the values in 'Participation' turn into floats
# sat_2017['Participation'] = sat_2017['Participation'].map(lambda x: float(x[:-1]))
# act_2017['Participation'] = act_2017['Participation'].map(lambda x: float(x[:-1]))

In [31]:
# Now they are floats
act_2017.head(3)

State  Participation  English  Math  Reading  Science  Composite
0  National           60.0     20.3  20.7     21.4     21.0       21.0
1   Alabama          100.0     18.9  18.4     19.7     19.4       19.2
2    Alaska           65.0     18.7  19.8     20.4     19.9       19.8

In [32]:
# Now they are floats
sat_2017.head(3)

State  Participation  Evidence-Based Reading and Writing  Math  Total
0  Alabama            5.0                                 593   572   1165
1   Alaska           38.0                                 547   533   1080
2  Arizona           30.0                                 563   553   1116

- Fix any individual values preventing other columns from being the appropriate type.

In [33]:
# Fixed the individual values in 4c

# This is how I fixes the individual values

# # This fixes the error of Maryland in the column 'math_sat' Changes 52 to 524
# sat_2017.at[20, 'Math'] = 524

# # This fixes the error for Maryland in the column 'science_act' Changes 2.3 to 23.2
# act_2017.at[21, 'Science']= 23.2

# # This fixes the error for Wyoming in the column 'composite' Changes 20.2x to 20.2
# act_2017.at[51, 'Composite']= 20.2

- Finish your data modifications by making sure the columns are now typed appropriately.

In [2]:
#code

- Display the data types again to confirm they are correct.

In [34]:
#Code: 
act_2017.dtypes

State             object
Participation    float64
English          float64
Math             float64
Reading          float64
Science          float64
Composite        float64
dtype: object

In [35]:
sat_2017.dtypes

State                                  object
Participation                         float64
Evidence-Based Reading and Writing      int64
Math                                    int64
Total                                   int64
dtype: object

#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [37]:
sat_2017.columns

Index(['State', 'Participation', 'Evidence-Based Reading and Writing', 'Math',
       'Total'],
      dtype='object')

In [36]:
act_2017.columns

Index(['State', 'Participation', 'English', 'Math', 'Reading', 'Science',
       'Composite'],
      dtype='object')

In [40]:
# Making the columns lowercase
sat_2017.columns = sat_2017.columns.str.lower()
sat_2017.columns

Index(['state', 'participation', 'evidence-based reading and writing', 'math',
       'total'],
      dtype='object')

In [43]:
# Making the columns lowercase
act_2017.columns = act_2017.columns.str.lower()
act_2017.columns

Index(['state', 'participation', 'english', 'math', 'reading', 'science',
       'composite'],
      dtype='object')

In [47]:
# Renaming the sat_2017 columns
sat_2017.rename(columns={'state': 'sat_state',
                         'participation': 'sat_participation',
                         'math': 'sat_math',
                         'total': 'sat_total'
                    }, inplace=True)

In [49]:
sat_2017.head(2)

sat_state  sat_participation  sat_evidence_based_reading_and_writing  \
0   Alabama                5.0                                     593   
1    Alaska               38.0                                     547   

   sat_math  sat_total  
0       572       1165  
1       533       1080

In [52]:
# Renaming the act_2017 columns
act_2017.rename(columns={'state': 'act_state',
                         'participation': 'act_participation',
                         'english': 'act_english',
                         'math': 'act_math',
                         'reading': 'act_reading',
                         'science': 'act_science',
                         'composite': 'act_composite'
                    }, inplace=True)

In [53]:
act_2017.head(2)

act_state  act_participation  act_english  act_math  act_reading  \
0  National               60.0         20.3      20.7         21.4   
1   Alabama              100.0         18.9      18.4         19.7   

   act_science  act_composite  
0         21.0           21.0  
1         19.4           19.2

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**state**|object|sat_2017|All 50 states and the District of Columbia|
|**2017_sat_participation**|float|sat_2017|The overall percentage of people taking the SAT in that state in 2017|
|**2017_sat_evidence_based_reading_and_writing**|int|sat_2017|The average score for the evidence-based reading and writing section of the SAT for each state |
|**2017_sat_math**|int|sat_2017|The average score for the math section of the SAT for each state|
|**2017_sat_total**|int|sat_2017|The overall average total score of the SAT for each state |
|**state**|object|act_2017|All 50 states, the District of Columbia, and the national average for each section|
|**2017_participation**|float|act_2017|The overall percentage of people taking the ACT in that state in 2017|
|**2017_act_english**|float|act_2017|The average score for the english section of the ACT for each state and the national average|
|**2017_act_math**|float|act_2017|The average score for the math section of the ACT for each state and the national average|
|**2017_act_reading**|float|act_2017|The average score for the reading section of the ACT for each state and the national average |
|**2017_act_science**|float|act_2017|The average score for the science section of the ACT for each state and the national average |
|**2017_act_composite**|float|act_2017|The average score on the ACT for each state and the national average score|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [57]:
act_2017.head()

act_state  act_participation  act_english  act_math  act_reading  \
0  National               60.0         20.3      20.7         21.4   
1   Alabama              100.0         18.9      18.4         19.7   
2    Alaska               65.0         18.7      19.8         20.4   
3   Arizona               62.0         18.6      19.8         20.1   
4  Arkansas              100.0         18.9      19.0         19.7   

   act_science  act_composite  
0         21.0           21.0  
1         19.4           19.2  
2         19.9           19.8  
3         19.8           19.7  
4         19.5           19.4

In [61]:
#code
# act_2017 contains an extra row

# This is the code to drop the extra row
# act_2017.drop([act_2017.index[0]], inplace=True)

In [102]:
act_2017.shape

(51, 7)

In [104]:
sat_2017.head(3)

sat_state  sat_participation  sat_evidence_based_reading_and_writing  \
0   Alabama                5.0                                     593   
1    Alaska               38.0                                     547   
2   Arizona               30.0                                     563   

   sat_math  sat_total  
0       572       1165  
1       533       1080  
2       553       1116

In [107]:
act_2017.rename(columns={'act_state': '2017_state'}, inplace=True)

In [110]:
sat_2017.rename(columns={'sat_state': '2017_state'}, inplace=True)

In [112]:
sat_2017.head(3)

2017_state  sat_participation  sat_evidence_based_reading_and_writing  \
0    Alabama                5.0                                     593   
1     Alaska               38.0                                     547   
2    Arizona               30.0                                     563   

   sat_math  sat_total  
0       572       1165  
1       533       1080  
2       553       1116

In [113]:
act_2017.head(3)

2017_state  act_participation  act_english  act_math  act_reading  \
1    Alabama              100.0         18.9      18.4         19.7   
2     Alaska               65.0         18.7      19.8         20.4   
3    Arizona               62.0         18.6      19.8         20.1   

   act_science  act_composite  
1         19.4           19.2  
2         19.9           19.8  
3         19.8           19.7

In [118]:
act_2017.columns

Index(['2017_state', 'act_participation', 'act_english', 'act_math',
       'act_reading', 'act_science', 'act_composite'],
      dtype='object')

In [119]:
act_2017.rename(columns={'act_participation': '2017_act_participation',
                         'act_english': '2017_act_english',
                         'act_math': '2017_act_math',
                         'act_reading': '2017_act_reading',
                         'act_science': '2017_act_science',
                         'act_composite': '2017_act_composite'}, inplace=True)

In [122]:
sat_2017.columns

Index(['2017_state', 'sat_participation',
       'sat_evidence_based_reading_and_writing', 'sat_math', 'sat_total'],
      dtype='object')

In [123]:
sat_2017.rename(columns={'sat_participation': '2017_sat_participation',
                         'sat_evidence_based_reading_and_writing': '2017_sat_evidence_based_reading_and_writing',
                         'sat_math': '2017_sat_math',
                         'sat_total': '2017_sat_total'}, inplace=True)

In [124]:
sat_2017.head()

2017_state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  
0                                          593            572            1165  
1                                          547            533            1080  
2                                          563            553            1116  
3                                          614            594            1208  
4                                          531            524            1055

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [125]:
#Code: 
# This code will merge the dataframe
#pd.merge(pd.merge(openprice,stockname,on='Symbol')
combined_2017 = pd.merge(sat_2017, act_2017, on='2017_state')
combined_2017.head()

2017_state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   
1                                          547            533            1080   
2                                          563            553            1116   
3                                          614            594            1208   
4                                          531            524            1055   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   
1                    65.0              18.7           19.8              20.4   
2                    62.0              18.6           19.8              20.1   
3                   100.0              18.9           19.0              19.7   
4                    31.0              22.5           22.7              23.1   

   2017_act_science  2017_act_composite  
0              19.4                19.2  
1              19.9                19.8  
2              19.8                19.7  
3              19.5                19.4  
4              22.2                22.8

In [142]:
combined_2017.rename(columns={'2017_state': 'state'}, inplace=True)

In [143]:
combined_2017.head()

state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   
1                                          547            533            1080   
2                                          563            553            1116   
3                                          614            594            1208   
4                                          531            524            1055   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   
1                    65.0              18.7           19.8              20.4   
2                    62.0              18.6           19.8              20.1   
3                   100.0              18.9           19.0              19.7   
4                    31.0              22.5           22.7              23.1   

   2017_act_science  2017_act_composite  
0              19.4                19.2  
1              19.9                19.8  
2              19.8                19.7  
3              19.5                19.4  
4              22.2                22.8

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [145]:
#code
# report.to_csv('report.csv')
combined_2017.to_csv('combined_2017.csv')

## 2018 Data Import and Cleaning

Read in the `sat_2018.csv` and `act_2018.csv` files and assign them to appropriately named pandas dataframes. For the **2018 ACT Data**, only the `Composite` scores are available. Repeat the same processes to clean the 2018 data here as you were instructed in the previous sections above.

In [67]:
sat_2018 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/sat_2018.csv')
act_2018 = pd.read_csv('/Users/aidancurley/Documents/dsir/Submissions/Projects/project-1-master/data/act_2018.csv')

In [68]:
sat_2018.shape

(51, 5)

In [73]:
act_2018.shape

(52, 3)

In [74]:
# There are no values that seem out of place or not possible
sat_2018.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           563.686275  556.235294  1120.019608
std                             47.502627   47.772623    94.155083
min                            480.000000  480.000000   977.000000
25%                            534.500000  522.500000  1057.500000
50%                            552.000000  544.000000  1098.000000
75%                            610.500000  593.500000  1204.000000
max                            643.000000  655.000000  1298.000000

In [75]:
# There are no values that seem out of place or not possible
act_2018.describe()

Composite
count  52.000000
mean   21.544231
std     2.119417
min    17.700000
25%    19.975000
50%    21.300000
75%    23.725000
max    25.600000

In [76]:
# Participation has to become a float
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [77]:
# Participation has to become a float
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   Composite      52 non-null     float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


In [78]:
# Updating the values in the 'Participation' column
sat_2018['Participation'] = sat_2018['Participation'].map(lambda x: float(x[:-1]))

In [79]:
# Updating the values in the 'Participation' column
act_2018['Participation'] = act_2018['Participation'].map(lambda x: float(x[:-1]))

In [82]:
# Now I need to get rid of the error
# There is an extra row in the act_2018 dataset. Maine was put in twice
act_2018.shape

(52, 3)

In [83]:
act_2018 = act_2018.drop_duplicates()

In [86]:
# Now it has the appropriate amount of columns
act_2018.shape

(51, 3)

In [87]:
sat_2018.columns

Index(['State', 'Participation', 'Evidence-Based Reading and Writing', 'Math',
       'Total'],
      dtype='object')

In [88]:
sat_2018.columns = sat_2018.columns.str.lower()
sat_2018.columns

Index(['state', 'participation', 'evidence-based reading and writing', 'math',
       'total'],
      dtype='object')

In [90]:
# Renaming the sat_2018 columns
sat_2018.rename(columns={'state': 'sat_state',
                         'participation': 'sat_participation',
                         'evidence-based reading and writing': 'sat_evidence_based_reading_and_writing',
                         'math': 'sat_math',
                         'total': 'sat_total'
                    }, inplace=True)

In [92]:
sat_2018.head(1)

sat_state  sat_participation  sat_evidence_based_reading_and_writing  \
0   Alabama                6.0                                     595   

   sat_math  sat_total  
0       571       1166

In [95]:
act_2018.columns = act_2018.columns.str.lower()
act_2018.columns

Index(['state', 'participation', 'composite'], dtype='object')

In [96]:
act_2018.rename(columns={'state': 'act_state',
                         'participation': 'act_participation',
                         'composite': 'act_composite'
                    }, inplace=True)

/Users/aidancurley/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [101]:
act_2018.head(3)

act_state  act_participation  act_composite
0   Alabama              100.0           19.1
1    Alaska               33.0           20.8
2   Arizona               66.0           19.2

In [127]:
act_2018.columns

Index(['act_state', 'act_participation', 'act_composite'], dtype='object')

In [134]:
act_2018.rename(columns={'act_state': '2018_state',
                         'act_participation': '2018_act_participation',
                         'act_composite': '2018_act_composite'
                    }, inplace=True)

/Users/aidancurley/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [135]:
act_2018.head()

2018_state  2018_act_participation  2018_act_composite
0     Alabama                   100.0                19.1
1      Alaska                    33.0                20.8
2     Arizona                    66.0                19.2
3    Arkansas                   100.0                19.4
4  California                    27.0                22.7

In [131]:
sat_2018.rename(columns={'sat_state': '2018_state',
                         'sat_participation': '2018_sat_participation',
                         'sat_evidence_based_reading_and_writing': '2018_sat_evidence_based_reading_and_writing',
                         'sat_math': '2018_sat_math',
                         'sat_total': '2018_sat_total'
                    }, inplace=True)

In [132]:
sat_2018.head()

2018_state  2018_sat_participation  \
0     Alabama                     6.0   
1      Alaska                    43.0   
2     Arizona                    29.0   
3    Arkansas                     5.0   
4  California                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  
0                                          595            571            1166  
1                                          562            544            1106  
2                                          577            572            1149  
3                                          592            576            1169  
4                                          540            536            1076

In [136]:
combined_2018 = pd.merge(sat_2018, act_2018, on='2018_state')
combined_2018.head()

2018_state  2018_sat_participation  \
0     Alabama                     6.0   
1      Alaska                    43.0   
2     Arizona                    29.0   
3    Arkansas                     5.0   
4  California                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   
1                                          562            544            1106   
2                                          577            572            1149   
3                                          592            576            1169   
4                                          540            536            1076   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

In [137]:
combined_2018.columns

Index(['2018_state', '2018_sat_participation',
       '2018_sat_evidence_based_reading_and_writing', '2018_sat_math',
       '2018_sat_total', '2018_act_participation', '2018_act_composite'],
      dtype='object')

In [139]:
combined_2018.rename(columns={'2018_state': 'state'}, inplace=True)

In [140]:
combined_2018.head()

state  2018_sat_participation  \
0     Alabama                     6.0   
1      Alaska                    43.0   
2     Arizona                    29.0   
3    Arkansas                     5.0   
4  California                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   
1                                          562            544            1106   
2                                          577            572            1149   
3                                          592            576            1169   
4                                          540            536            1076   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

In [146]:
combined_2018.to_csv('combined_2018.csv')

# Data dictionary for ACT and SAT data in 2018
|Feature|Type|Dataset|Description|
|---|---|---|---|
|**state**|object|sat_2018|All 50 states and the District of Columbia|
|**2018_sat_participation**|float|sat_2018|The overall percentage of people taking the SAT in that state in 2018|
|**2018_sat_evidence_based_reading_and_writing**|int|sat_2018|The average score for the evidence-based reading and writing section of the SAT for each state |
|**2018_sat_math**|int|sat_2018|The average score for the math section of the SAT for each state|
|**2018_sat_total**|int|sat_2018|The overall average total score of the SAT for each state |
|**state**|object|act_2018|All 50 states and the District of Columbia|
|**2018_act_participation**|float|act_2018|The overall percentage of people taking the ACT in that state in 2018|
|**2018_act_composite**|float|act_2018|The average score on the ACT for each state and the national average score|

In [148]:
combined_2017.head(1)

state  2017_sat_participation  \
0  Alabama                     5.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   

   2017_act_science  2017_act_composite  
0              19.4                19.2

In [149]:
combined_2018.head(1)

state  2018_sat_participation  \
0  Alabama                     6.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [151]:
combined_2017.head()

state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   
1                                          547            533            1080   
2                                          563            553            1116   
3                                          614            594            1208   
4                                          531            524            1055   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   
1                    65.0              18.7           19.8              20.4   
2                    62.0              18.6           19.8              20.1   
3                   100.0              18.9           19.0              19.7   
4                    31.0              22.5           22.7              23.1   

   2017_act_science  2017_act_composite  
0              19.4                19.2  
1              19.9                19.8  
2              19.8                19.7  
3              19.5                19.4  
4              22.2                22.8

In [153]:
combined_2018.head()

state  2018_sat_participation  \
0     Alabama                     6.0   
1      Alaska                    43.0   
2     Arizona                    29.0   
3    Arkansas                     5.0   
4  California                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   
1                                          562            544            1106   
2                                          577            572            1149   
3                                          592            576            1169   
4                                          540            536            1076   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

In [156]:
final = pd.merge(combined_2017, combined_2018, on='state')
final.head()

state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   
1                                          547            533            1080   
2                                          563            553            1116   
3                                          614            594            1208   
4                                          531            524            1055   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   
1                    65.0              18.7           19.8              20.4   
2                    62.0              18.6           19.8              20.1   
3                   100.0              18.9           19.0              19.7   
4                    31.0              22.5           22.7              23.1   

   2017_act_science  2017_act_composite  2018_sat_participation  \
0              19.4                19.2                     6.0   
1              19.9                19.8                    43.0   
2              19.8                19.7                    29.0   
3              19.5                19.4                     5.0   
4              22.2                22.8                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   
1                                          562            544            1106   
2                                          577            572            1149   
3                                          592            576            1169   
4                                          540            536            1076   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

In [157]:
final.to_csv('final.csv')

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [159]:
#Code:
final.describe().T

count      mean        std  \
2017_sat_participation                        50.0    38.600  34.560211   
2017_sat_evidence_based_reading_and_writing   50.0   570.860  44.385146   
2017_sat_math                                 50.0   558.660  45.839832   
2017_sat_total                                50.0  1129.620  89.913085   
2017_act_participation                        50.0    65.920  32.110676   
2017_act_english                              50.0    20.862   2.324307   
2017_act_math                                 50.0    21.136   1.973990   
2017_act_reading                              50.0    21.956   2.046317   
2017_act_science                              50.0    21.410   1.731962   
2017_act_composite                            50.0    21.466   2.004241   
2018_sat_participation                        50.0    44.820  37.097604   
2018_sat_evidence_based_reading_and_writing   50.0   565.020  47.010416   
2018_sat_math                                 50.0   557.760  46.987432   
2018_sat_total                                50.0  1122.880  92.845634   
2018_act_participation                        50.0    62.320  34.114118   
2018_act_composite                            50.0    21.454   2.111312   

                                               min       25%      50%  \
2017_sat_participation                         2.0     4.000    34.00   
2017_sat_evidence_based_reading_and_writing  503.0   535.250   559.50   
2017_sat_math                                492.0   524.000   549.50   
2017_sat_total                               996.0  1057.000  1107.50   
2017_act_participation                         8.0    31.000    71.00   
2017_act_english                              16.3    19.000    20.55   
2017_act_math                                 18.0    19.400    20.90   
2017_act_reading                              18.1    20.425    21.70   
2017_act_science                              18.2    19.925    21.30   
2017_act_composite                            17.8    19.800    21.40   
2018_sat_participation                         2.0     4.250    50.00   
2018_sat_evidence_based_reading_and_writing  480.0   535.000   552.00   
2018_sat_math                                486.0   524.250   545.50   
2018_sat_total                               998.0  1062.250  1098.50   
2018_act_participation                         7.0    27.750    66.00   
2018_act_composite                            17.7    19.925    21.05   

                                                  75%     max  
2017_sat_participation                         65.000   100.0  
2017_sat_evidence_based_reading_and_writing   613.500   644.0  
2017_sat_math                                 601.000   651.0  
2017_sat_total                               1214.000  1295.0  
2017_act_participation                        100.000   100.0  
2017_act_english                               23.100    25.5  
2017_act_math                                  23.000    25.3  
2017_act_reading                               23.875    26.0  
2017_act_science                               22.975    24.9  
2017_act_composite                             23.400    25.5  
2018_sat_participation                         74.500   100.0  
2018_sat_evidence_based_reading_and_writing   612.750   643.0  
2018_sat_math                                 594.250   655.0  
2018_sat_total                               1207.000  1298.0  
2018_act_participation                        100.000   100.0  
2018_act_composite                             23.650    25.6

In [177]:
final.head()

state  2017_sat_participation  \
0     Alabama                     5.0   
1      Alaska                    38.0   
2     Arizona                    30.0   
3    Arkansas                     3.0   
4  California                    53.0   

   2017_sat_evidence_based_reading_and_writing  2017_sat_math  2017_sat_total  \
0                                          593            572            1165   
1                                          547            533            1080   
2                                          563            553            1116   
3                                          614            594            1208   
4                                          531            524            1055   

   2017_act_participation  2017_act_english  2017_act_math  2017_act_reading  \
0                   100.0              18.9           18.4              19.7   
1                    65.0              18.7           19.8              20.4   
2                    62.0              18.6           19.8              20.1   
3                   100.0              18.9           19.0              19.7   
4                    31.0              22.5           22.7              23.1   

   2017_act_science  2017_act_composite  2018_sat_participation  \
0              19.4                19.2                     6.0   
1              19.9                19.8                    43.0   
2              19.8                19.7                    29.0   
3              19.5                19.4                     5.0   
4              22.2                22.8                    60.0   

   2018_sat_evidence_based_reading_and_writing  2018_sat_math  2018_sat_total  \
0                                          595            571            1166   
1                                          562            544            1106   
2                                          577            572            1149   
3                                          592            576            1169   
4                                          540            536            1076   

   2018_act_participation  2018_act_composite  
0                   100.0                19.1  
1                    33.0                20.8  
2                    66.0                19.2  
3                   100.0                19.4  
4                    27.0                22.7

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [167]:
from math import sqrt

In [174]:
# https://codereview.stackexchange.com/questions/9222/calculating-population-standard-deviation/9224

# This gets the mean
def mean(lst):
    return sum(lst) / len(lst)

def stddev(lst):
    """returns the standard deviation of lst"""
    variance = 0
    mn = mean(lst)
    for e in lst:
        variance += (e-mn)**2
    variance /= len(lst)

    return sqrt(variance)

numbers = [120,112,131,211,312,90]

print(stddev(numbers))

76.71085682975752


In [ ]:
# fahrenheit = {'t1': -30,'t2': -20,'t3': -10,'t4': 0}

# # Get the corresponding `celsius` values and create the new dictionary
# celsius = {k:(float(5)/9)*(v-32) for (k,v) in fahrenheit.items()}

In [182]:
final.keys()

Index(['state', '2017_sat_participation',
       '2017_sat_evidence_based_reading_and_writing', '2017_sat_math',
       '2017_sat_total', '2017_act_participation', '2017_act_english',
       '2017_act_math', '2017_act_reading', '2017_act_science',
       '2017_act_composite', '2018_sat_participation',
       '2018_sat_evidence_based_reading_and_writing', '2018_sat_math',
       '2018_sat_total', '2018_act_participation', '2018_act_composite'],
      dtype='object')

In [176]:
#code
final.items()
{k: stddev(v) for (k,v) in }

<generator object DataFrame.items at 0x119f21950>

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [ ]:
#Code:

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

Answer

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [2]:
#code

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

## Bonus: Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code: